### PR TITLE
CI efficiency improvements

### PR DESCRIPTION
**Pull request by achamayou**
This is trying to collapse our CI into something more compact, so that it:

1. runs faster (ie. less than 5 minutes)
2. is easier to make changes to (ie. single common.yml with switches)
3. is cheaper (some stuff doesn't need custom VMs)

Some of that is achieved by pushing expensive tasks off to #832 (SAN, Coverage).

### CODE REVIEW

File: .azure-pipelines-templates/matrix.yml [hunk 1]
```diff
@@ -1,188 +1,83 @@
...
-        ctest_args: '-LE "benchmark|perf|raft"'
-      CFT:
-        ctest_args: '-LE "benchmark|perf|pbft"'
+      ctest_args: '-LE "benchmark|perf" -j2'
     Release:
-      BFT:
-        ctest_args: '-LE "benchmark|perf|raft"'
-      CFT:
-        ctest_args: '-LE "benchmark|perf|pbft"'
+      ctest_args: '-LE "benchmark|perf" -j2'
```
Comment (eddyashton) [inline .azure-pipelines-templates/matrix.yml]: If we're doing a release, should we just run the full test suite?

File: .azure-pipelines-templates/matrix.yml [hunk 2]
```diff
@@ -1,188 +1,83 @@
...
-      dependsOn: ['Test_SGX_CFT_Release', 'Test_SGX_BFT_Release']
     Perf:
       pool: CCF-Perf
 
   build:
     common:
       cmake_args: '-DCMAKE_C_COMPILER_LAUNCHER="ccache" -DCMAKE_CXX_COMPILER_LAUNCHER="ccache"'
     NoSGX:
-      cmake_args: '-DCOMPILE_TARGETS=virtual -DCOVERAGE=ON'
+      cmake_args: '-DTARGET=virtual'
```
Comment (eddyashton) [inline .azure-pipelines-templates/matrix.yml]: Looks like a bad merge - this should still be `COMPILE_TARGETS`

File: .azure-pipelines-templates/matrix.yml (hunk 1 above)
Comment (achamayou) [reply .azure-pipelines-templates/matrix.yml]: Sure, but unless it's also Perf, we don't record the values. Seems fair though.

File: .azure-pipelines-templates/matrix.yml (hunk 2 above)
Comment (achamayou) [reply .azure-pipelines-templates/matrix.yml]: Yep, it was. Also the SGX build was really both (oops!).

**Comment by achamayou** [comment]
So as far as the tests go, there seem to be a number of prime candidates:

SGX:
image

No SGX:
image

File: .azure-pipelines-templates/matrix.yml [hunk 3]
```diff
@@ -1,188 +1,83 @@
...
+      cmake_args: '-DCOMPILE_TARGETS=virtual'
     SGX:
-      cmake_args: ''
+      cmake_args: '-DCOMPILE_TARGETS=sgx'
     debug:
-      cmake_args: '-DCMAKE_BUILD_TYPE=Debug -DVERBOSE_LOGGING=ON -DBUILD_SMALLBANK=OFF'
+      cmake_args: '-DCMAKE_BUILD_TYPE=Debug -DBUILD_SMALLBANK=OFF'
     install:
       install_prefix: '/tmp/ccf-install'
       cmake_args: '-DCMAKE_INSTALL_PREFIX=/tmp/ccf-install'
```
Comment (olgavrou) [inline .azure-pipelines-templates/matrix.yml]: SAN is moving to the nightly build?

Comment (achamayou) [reply .azure-pipelines-templates/matrix.yml]: Yep #832

File: .daily.yml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+schedules:
```
Comment (achamayou) [inline .daily.yml]: Can't set up the pipeline from a fork, annoyingly, so it'll have to get merged first.

**Comment by achamayou** [comment]
Added files for Daily build, but can't set up from a fork, so I will set up the pipeline once this merges.

File: src/node/rpc/test/member_voting_test.cpp [hunk 5]
```diff
@@ -821,7 +821,7 @@ DOCTEST_TEST_CASE("Propose raw writes")
   logger::config::level() = logger::INFO;
   DOCTEST_SUBCASE("insensitive tables")
   {
-    const auto n_members = 10;
+    const auto n_members = 3;
```
Comment (achamayou) [inline src/node/rpc/test/member_voting_test.cpp]: Incremental store setup is kinda expensive, ~10 seconds in all for this test. Down to 3 members, it's consistently <1s, and I don't think it changes what we're testing functionally here.

Once this is done, ctest -L unit is consistently <3s, and there's no point using -j anymore. Small win, but still a win.

File: cmake/common.cmake [hunk 6]
```diff
@@ -405,11 +405,10 @@ function(add_e2e_test)
       TEST ${PARSED_ARGS_NAME} APPEND
       PROPERTY ENVIRONMENT "PYTHONPATH=${CCF_DIR}/tests:$ENV{PYTHONPATH}"
     )
-    if(${PARSED_ARGS_IS_SUITE})
-      set_property(TEST ${PARSED_ARGS_NAME} APPEND PROPERTY LABELS suite)
-    else()
-      set_property(TEST ${PARSED_ARGS_NAME} APPEND PROPERTY LABELS end_to_end)
-    endif()
+    set_property(TEST ${PARSED_ARGS_NAME} APPEND PROPERTY LABELS end_to_end)
```
Comment (achamayou) [inline cmake/common.cmake]: This is to give us the freedom to split up stuff that doesn't parallelise well, like tests that do a lot of quote verification.

**Comment by achamayou** [comment]
Current builds are about 20 minutes, Pipelines doesn't have a good way to show that, this is the best I can do:

image

After this change we are in the same ballpark:

image

But we run much fewer VMs, and they're less than fully busy, so we should be able to get more gains from running some tests concurrently without paying minutes for packing/unpacking. ccache should also be warmer in general. This also shaves about 20% of our YAML off, even after adding the Daily build.

To be followed by more test optimisations...

File: .azure-pipelines-templates/daily-matrix.yml [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+      cmake_args: '-DCOMPILE_TARGETS=virtual'
+    daily:
+      cmake_args: '-DCMAKE_BUILD_TYPE=Debug -DSAN=ON -DCOVERAGE=ON -DBUILD_SMALLBANK=OFF'
+
+jobs:
+  - template: common.yml
+    parameters:
+      target: NoSGX
+      env: parameters.env.NoSGX
+      cmake_args: '${{ parameters.build.common.daily }} ${{ parameters.build.NoSGX.cmake_args }}'
```
Comment (eddyashton) [inline .azure-pipelines-templates/daily-matrix.yml]: Typo, first should be `parameters.build.daily.cmake_args`?

Comment (achamayou) [reply .azure-pipelines-templates/daily-matrix.yml]: Absolutely, thank you!

**Pull request merged**